alchemiscale_fah/compute/service.py

```diff
@@ -4,6 +4,621 @@
 
 """
 
+import os
+import asyncio
+import gc
+import json
+from typing import Union, Optional, List, Dict, Tuple
+from pathlib import Path
+from uuid import uuid4
+import threading
+import time
+import logging
+import shutil
+from concurrent.futures import ProcessPoolExecutor
 
-class FAHComputeService:
-    ...
+from gufe.tokenization import GufeKey, JSON_HANDLER
+from gufe.protocols.protocoldag import (
+    ProtocolDAG,
+    ProtocolDAGResult,
+    _pu_to_pur,
+    Context,
+)
+from gufe.protocols.protocolunit import ProtocolUnit, ProtocolUnitResult
+
+from alchemiscale.models import Scope, ScopedKey
+from alchemiscale.storage.models import Task, TaskHub, ComputeServiceID
+from alchemiscale.compute.client import AlchemiscaleComputeClient
+from alchemiscale.compute.service import (
+    SynchronousComputeService,
+    InterruptableSleep,
+    SleepInterrupted,
+)
+from alchemiscale.keyedchain import KeyedChain
+
+from .models import FahProject
+from .settings import FahAsynchronousComputeServiceSettings
+from .client import FahAdaptiveSamplingClient
+from .index import FahComputeServiceIndex
+from ..protocols.protocolunit import FahSimulationUnit, FahContext
+
+# NOTE: if we don't do this, it appears that our RDKit mols won't keep their
+# properties on being pickled with a ProcessPoolExecutor, despite our efforts to avoid this;
+# this is important for retaining partial charges
+from rdkit import Chem
+
+Chem.SetDefaultPickleProperties(Chem.PropertyPickleOptions.AllProps)
+
+
+class FahAsynchronousComputeService(SynchronousComputeService):
+    """An asynchronous compute service for utilizing a Folding@Home work server."""
+
+    def __init__(self, settings: FahAsynchronousComputeServiceSettings):
+        """Create a `FAHSynchronousComputeService` instance."""
+        self.settings = settings
```

Comment:
Either unpack fully or read from settings object, having both allows state to get out of sync between the objects (even if one isn't mutable.)

Reply:
See above on subclassing `SynchronousComputeService`; will need to make change there to allow us to snip these as attributes.